api/provider/aws/service_list.go

```diff
@@ -1,110 +1,81 @@
-	}
-
-	serviceSummaries := []models.ServiceSummary{}
-	for _, serviceArns := range clusterServices {
-		for _, serviceArn := range serviceArns {
-			ecsServiceID := s.serviceARNToECSServiceID(aws.StringValue(serviceArn))
-			service := &models.Service{
-				ServiceID: ecsServiceID,
-			}
-			s.populateSummariesTags(service, ecsServiceID)
-
-			summary := models.ServiceSummary{
-				ServiceID:       service.ServiceID,
-				ServiceName:     service.ServiceName,
-				EnvironmentID:   service.EnvironmentID,
-				EnvironmentName: service.EnvironmentName,
-			}
-
-			serviceSummaries = append(serviceSummaries, summary)
-		}
+		serviceIDs[i] = serviceID
 	}
 
-	return serviceSummaries, nil
+	return s.makeServiceSummaryModels(serviceIDs)
 }
 
-func (s *ServiceProvider) listServices(clusterArn *string) ([]*string, error) {
-	serviceArns := []*string{}
-	input := &ecs.ListServicesInput{
-		Cluster: clusterArn,
-	}
-
-	for {
-		output, err := s.AWS.ECS.ListServices(input)
-		if err != nil {
-			return nil, err
+func (s *ServiceProvider) listServiceARNs() ([]string, error) {
+	var serviceARNs []string
+	fn := func(output *ecs.ListServicesOutput, lastPage bool) bool {
+		for _, serviceARN := range output.ServiceArns {
+			serviceARNs = append(serviceARNs, aws.StringValue(serviceARN))
 		}
 
-		serviceArns = append(serviceArns, output.ServiceArns...)
-		input.NextToken = output.NextToken
+		return !lastPage
+	}
 
-		if output.NextToken == nil {
-			break
-		}
+	if err := s.AWS.ECS.ListServicesPages(&ecs.ListServicesInput{}, fn); err != nil {
+		return nil, err
 	}
 
-	return serviceArns, nil
+	return serviceARNs, nil
 }
 
-func (s *ServiceProvider) listClusters() ([]*string, error) {
-	input := &ecs.ListClustersInput{}
-	clusterArns := []*string{}
+func (s *ServiceProvider) makeServiceSummaryModels(serviceIDs []string) ([]models.ServiceSummary, error) {
+	serviceTags, err := s.TagStore.SelectByType("service")
+	if err != nil {
+		return nil, err
+	}
+
+	models := make([]models.ServiceSummary, len(serviceIDs))
+	for i, serviceID := range serviceIDs {
+		models[i].ServiceID = serviceID
+
+		if tag, ok := serviceTags.WithID(serviceID).WithKey("name").First(); ok {
+			models[i].ServiceName = tag.Value
+		}
 
-	for {
-		output, err := s.AWS.ECS.ListClusters(input)
+		environmentID, err := lookupEntityEnvironmentID(s.TagStore, "service", serviceID)
 		if err != nil {
 			return nil, err
 		}
 
-		clusterArns = append(clusterArns, output.ClusterArns...)
-		input.NextToken = output.NextToken
+		models[i].EnvironmentID = environmentID
 
-		if output.NextToken == nil {
-			break
+		environmentTags, err := s.TagStore.SelectByTypeAndID("environment", environmentID)
+		if err != nil {
+			return nil, err
 		}
-	}
 
-	return clusterArns, nil
-}
+		tag, ok := environmentTags.WithKey("name").First()
+		if !ok {
+			return nil, errors.Newf(errors.EnvironmentDoesNotExist, "Could not resolve name for environment with id '%s'", environmentID)
```

Comment:
This does not mean the environment does not exist. We don't error when we can't find name tags.

Reply:
✅